// @sudoku/stores/domain.js
import { writable } from 'svelte/store';
import { createGame, createSudoku } from '../../../domain/index.js';

// 初始化一个 9x9 全空白数组
const emptyGrid = Array(9).fill(0).map(() => Array(9).fill(0));

function createDomainStore() {
    // 暴露出响应式状态，包含 UI 需要的所有变量
    const { subscribe, set } = writable({
        grid: emptyGrid,
        initialGrid: emptyGrid,
        invalidCells: [],
        isSolved: false,
        canUndo: false,
        canRedo: false
    });

    let gameInstance = null;

    // 每次领域发生变化，就将状态推入响应式 writable
    function sync() {
        if (!gameInstance) return;
        const currentSudoku = gameInstance.getSudoku();
        const initialSudoku = gameInstance.getInitialSudoku();

        set({
            grid: currentSudoku.getGrid(),
            initialGrid: initialSudoku.getGrid(),
            // 组件原本用的是 ['x,y'] 字符串，我们适配成同样的格式
            invalidCells: currentSudoku.getInvalidCells().map(cell => `${cell.x},${cell.y}`),
            isSolved: currentSudoku.isSolved(),
            canUndo: gameInstance.canUndo(),
            canRedo: gameInstance.canRedo()
        });
    }

    // 默认给个空
    gameInstance = createGame({ sudoku: createSudoku(emptyGrid) });
    sync();

    return {
        subscribe,

        // 游戏初始化，在 UI 层调用 grid.generate 或 grid.decodeSencode 后会调用这里的 init 来创建新的游戏实例
        init(initialGridData) {
            gameInstance = createGame({ sudoku: createSudoku(initialGridData) });
            sync();
        },

        // 供 UI 下子
        guess(row, col, value) {
            if (gameInstance) {
                gameInstance.guess({ row, col, value });
                sync();
            }
        },

        // 使用提示并将其固化
        applyHint(row, col, value) {
            if (gameInstance) {
                gameInstance.applyHint({ row, col, value });
                sync();
            }
        },

        undo() {
            if (gameInstance) {
                gameInstance.undo();
                sync();
            }
        },

        redo() {
            if (gameInstance) {
                gameInstance.redo();
                sync();
            }
        },

        // 把 domain 对象本身也暴露出去，灵活性更高
        getGame() {
            return gameInstance;
        }
    };
}

// 导出store
export const domainStore = createDomainStore();
